service/src/main/java/bio/terra/cbas/dependencies/bard/BardServiceImpl.java

```diff
@@ -0,0 +1,91 @@
+package bio.terra.cbas.dependencies.bard;
+
+import bio.terra.bard.api.DefaultApi;
+import bio.terra.bard.client.ApiClient;
+import bio.terra.bard.model.EventsEventLogRequest;
+import bio.terra.cbas.config.BardServerConfiguration;
+import bio.terra.cbas.dependencies.common.HealthCheck;
+import bio.terra.cbas.model.RunSetRequest;
+import bio.terra.cbas.models.GithubMethodDetails;
+import bio.terra.cbas.models.MethodVersion;
+import bio.terra.common.iam.BearerToken;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Component;
+
+@Component
+public class BardServiceImpl implements BardService, HealthCheck {
+
+  Logger log = LoggerFactory.getLogger(BardServiceImpl.class);
+  private static final String appId = "cbas";
+  private final BardClient bardClient;
+  private final BardServerConfiguration bardServerConfiguration;
+
+  public BardServiceImpl(BardClient bardClient, BardServerConfiguration bardServerConfiguration) {
+    this.bardClient = bardClient;
+    this.bardServerConfiguration = bardServerConfiguration;
+  }
+
+  private DefaultApi getDefaultAuthApi(BearerToken userToken) {
+    ApiClient client = bardClient.bardAuthClient(userToken);
+    return bardClient.defaultApi(client);
+  }
+
+  @Override
+  public void logRunSetEvent(
+      RunSetRequest request,
+      MethodVersion methodVersion,
+      List<String> workflowIds,
+      BearerToken userToken) {
+    String eventName = "workflow-submission";
+    HashMap<String, String> properties = new HashMap<>();
+    properties.put("runSetName", request.getRunSetName());
+    properties.put("methodName", methodVersion.method().name());
+    properties.put("methodSource", methodVersion.method().methodSource());
+    properties.put("methodVersionName", methodVersion.name());
+    properties.put("methodVersionUrl", methodVersion.url());
+    properties.put("recordCount", String.valueOf(request.getWdsRecords().getRecordIds().size()));
+    properties.put("workflowIds", workflowIds.toString());
+
+    Optional<GithubMethodDetails> maybeGitHubMethodDetails =
+        methodVersion.method().githubMethodDetails();
+    if (maybeGitHubMethodDetails.isPresent()) {
+      GithubMethodDetails githubMethodDetails = maybeGitHubMethodDetails.get();
+      properties.put("githubOrganization", githubMethodDetails.organization());
+      properties.put("githubRepository", githubMethodDetails.repository());
+      properties.put("githubIsPrivate", githubMethodDetails.isPrivate().toString());
+    }
+
+    logEvent(eventName, properties, userToken);
+  }
+
+  @Async("bardAsyncExecutor")
+  @Override
+  public void logEvent(String eventName, Map<String, String> properties, BearerToken userToken) {
```

Comment:
Note that in Spring, the `Async` annotation only applies to proxied calls. My mental model is something like:

- Class: "I want an autowired BardService"
- Spring: "Ah yes, I can make one of these BardServiceImpls. Oh, and it has async methods so I'm going to create a proxied version which turns this method into one that returns a Future"
- Class: (receives a new, proxying version of the BardServiceImpl)

So in this case, the `logEvent` will only be async if it's called from an autowired reference to BardService. And because this one is coming from _inside_ the same object, it won't be made asynchronously here.

Reply:
Incidentally, I think this is also why the test were having errors referencing "that's a proxy not a BardService instance", and making the tests reference the interface makes the mocking work

Reply:
In that case could I get rid of the `BardService` interface?

Reply:
I tried getting rid of the interface and got the proxy error again btw